### ISSUE TITLE
 Update code for compatibility with PHP8

### ISSUE THREAD

**Issue by JimBacon** [Indicia-Team/warehouse#366]
Relates to https://github.com/BiologicalRecordsCentre/iRecord/issues/1047

Also update check_php_version() in modules/indicia_setup/helpers/config_test.php and the installation docs at https://indicia-docs.readthedocs.io/en/latest/administrating/warehouse/warehouse-installation.html

**Comment by JimBacon** [comment]
Running PHP Compatibility checks for PHP >= 7.3 returns the following summary results

```
-----------------------------------------------------------------------------------------
FILE                                                                    ERRORS  WARNINGS
-----------------------------------------------------------------------------------------
occurrence_comment_quick_reply_page.php                                     0       1
application/helpers/vague_date.php                                          0       1
application/libraries/MY_Database.php                                       0       1
application/libraries/WorkQueue.php                                         0       1
application/models/taxa_taxon_list.php                                      0       1
application/models/termlists_term.php                                       0       1
modules/indicia_svc_data/views/services/data/termlists_term/li.php          0       1
system/config/encryption.php                                                2       0
system/core/Kohana.php                                                      23      0
system/core/utf8.php                                                        0       1
system/core/utf8/ucwords.php                                                1       0
system/libraries/Database.php                                               0       1
system/libraries/Encrypt.php                                                15      0
system/libraries/Input.php                                                  0       2
system/libraries/ORM.php                                                    0       1
system/libraries/Validation.php                                             0       2
system/libraries/drivers/Cache/Sqlite.php                                   5       0
system/libraries/drivers/Database/Mssql.php                                 15      0
system/libraries/drivers/Database/Mysql.php                                 15      0
system/libraries/drivers/Database/Pdosqlite.php                             1       0
system/vendor/swift/Swift/Connection/Sendmail.php                           0       1
system/vendor/swift/Swift/Connection/SMTP.php                               0       2
system/vendor/swift/Swift/Iterator/MySQLResult.php                          5       0
system/vendor/swift/Swift/Message/Encoder.php                               0       4
system/vendor/swift/Swift/Plugin/MailSend.php                               0       1
vendor/maennchen/zipstream-php/test/BigintTest.php                          4       0
```
The errors are all in the Kohana system code

**Comment by JimBacon** [comment]
Current situation is that I am stuck on unit testing for PHP8 as we have a dependency on DbUnit which is obsolete.
See https://github.com/Indicia-Team/warehouse/issues/375#issuecomment-873594919

**Comment by johnvanbreda** [comment]
Although I don't like this as a solution, do we need a re-written set of tests that work in recent PHPUnit versions?

**Comment by JimBacon** [comment]
I have modified all the tests so that they are working with the latest version of PHPUnit.
That was necessary as we needed to use a version of PHPUnit compatible with PHP8.

The issue is with our modules/phpUnit/libraries/Indicia_DatabaseTestCase which inherits from PHPUnit\DbUnit\TestCase.
The original, https://github.com/sebastianbergmann/dbunit, is no longer maintained.
I have worked around this by employing a fork, https://github.com/misantron/dbunit.
It is not a fork that gives any indication of ongoing maintenance.
Installed with Composer, it is limited to PHP7.

A possible solution is to include the source code of DbUnit in our repo and keep it compatible with changes to PHPUnit.

However, the justification for DbUnit being abandonned was

> it is usually easier to implement a custom, project-specific solution for database fixture management than to use the generic solution provided by DbUnit. The effort to build such a custom solution is usually quite low, resulting in less than two hundred lines of code in a trait that is used by the database test case classes.

So that is the other possible solution.

**Comment by JimBacon** [comment]
I just pushed [some work](https://github.com/Indicia-Team/warehouse/commit/f81c13667761165239e60f30df9fc3e85761a493) I did a while back so it doesn't get lost. This is working towards replacing DbUnit by creating my own, very simple, drop-in alternative for creating fixtures.

All the current tests which manipulate the database would need updating to follow this pattern and then DbUnit could be discarded. That would enable us to switch to PHP8 without losing our ability to run the tests.

**Comment by johnvanbreda** [comment]
@jimbacon it might be helpful to know what changes are required to existing tests - is it only when fixtures are used to predefine data? If so then this may only affect a handful of tests (other than whatever is required to update the main fixtures).

**Comment by JimBacon** [comment]
It is just about tests where we set up a database with pre-defined data. I haven't got a clear idea of how much work that is but manageable, I imagine. I don't feel like I have necessarily understood all the things that DbUnit does but my first attempt to replace it seemed to work okay. 

To update the test
 - I changed the parent class to the new one I've created (SimpleDatabaseTestCase)
 - I created the fixture definition by merging two arrays, a core fixture used widely, and a local fixture specific to the test in hand.

The diff makes the changes look more significant that they were because there was a change of indent to the local fixture definition.

If you are developing any new tests, it would be great if you could move in this direction. Very happy for you to offer any improvements to what I have done. (Its all in the PHP8 branch at present but a merge with dev would probably not cause a problem.)

**Comment by JimBacon** [comment]
I just noticed that the master branch of https://github.com/misantron/dbunit does support PHP8 but it has not been released on packagist. By making use of this version, I have been able to run the unit tests with PHP8. That revealed some compatability issues which I fixed so all the tests pass.

I have merged the work in to the develop branch and set Travis to run the unit tests on both PHP7.3 and PHP8.0 henceforth.

Although the test coverage is limited, and there may still be a few compatibility issues lurking, I'm going to close this issue. It is probably a good idea to continue moving away from using DbUnit but it is no longer blocking us.
